client/sniffer/src/main/java/org/elasticsearch/client/sniff/ElasticsearchNodesSniffer.java

```diff
@@ -218,21 +224,75 @@ private static Node readNode(String nodeId, JsonParser parser, Scheme scheme) th
         if (publishedHost == null) {
             logger.debug("skipping node [" + nodeId + "] with http disabled");
             return null;
-        } else {
-            logger.trace("adding node [" + nodeId + "]");
-            if (version.startsWith("2.")) {
-                /*
-                 * 2.x doesn't send roles, instead we try to read them from
-                 * attributes.
-                 */
-                master = masterAttribute == null ? false == clientAttribute : masterAttribute;
-                data = dataAttribute == null ? false == clientAttribute : dataAttribute;
-            } else {
-                assert sawRoles : "didn't see roles for [" + nodeId + "]";
+        }
+
+        Map<String, List<String>> realAttributes = new HashMap<>(protoAttributes.size());
+        List<String> keys = new ArrayList<>(protoAttributes.keySet());
+        for (String key : keys) {
+            if (key.endsWith(".0")) {
```

Comment:
oh man we don't get the array but rather each item this way?

Reply:
Yes....